config/config.go

```diff
@@ -371,10 +372,22 @@ func GetGlobalConfig() *Config {
 
 // Load loads config options from a toml file.
 func (c *Config) Load(confFile string) error {
```

Comment:
When upgrade from old version tidb to a newer version, some old config item maybe is not used in the newer version tidb. The rolling upgrade may meet error in this case, then the rolling upgrade failed.

How about only check the config items when `-config-check` is provided?

Reply:
I don't like this suggestion. I think the behavior should be that tidb-server simply refuses to start if there are unknown configuration options, for the reasons I provided above.

A couple points:

1) Now that there's a `--config-check` option, it can be a part of the recommended procedure for a rolling upgrade, so that the only people running into the issue you describe are those *not* following the recommended procedure.

2) The benefit of a rolling upgrade is that you never have to take the entire cluster offline. When upgrading the first node, it will refuse to start with this error. That's a fine time to address the problem, apply the change to the config files, get that first node upgraded, and then proceed with the rolling upgrade.

@morgo, your thoughts?

Reply:
I think because our own CI broke, we may need to ease this change in slowly. Here would be my suggestion:

1. Keep `--config-check` as is, it is independently very useful.
2. Change from refuse to start to print a `WARN` about each incorrect config item.
3. Add a new flag for `--config-strict`, which defaults to `FALSE`.  We can document it that enabling it is a best practice, and we may change the default to `TRUE` in the future.  This will result in errors for invalid options.

It is not bulletproof of course, because the default log level is `INFO` it is possible that the warnings may not be seen.

Reply:
I think the very fact our own CI broke is evidence of the value of this behavior change. Even *we* weren't keeping track of whether our config files were well-formed!

2. Your suggestion above is difficult, because the logging system isn't set up yet when the config file is parsed. Either we have to keep around the list of undecoded items and check them later after logging is possible, or we have to pass some other string/structure/flag around, or we have to simply print something to stderr, or ...? @coocood Morgan mentioned that you may have some thoughts on this?

3. I dislike this. Adding this new flag means that people may start relying on it, and then we have to keep it around as a no-op if we later make that behavior the default.

Reply:
I do agree with Morgan's first and second suggestions. Our users may use old version ansible that doesn't pre-check the config using `tidb-server --config-check`, so we need to keep the cluster stable when rolling upgrade.

Reply:
Alright, it sounds like there's a consensus. Can someone offer a suggestion of the best way to technically implement a "warning" before the logging system is set up?

Reply:
@kolbe I will create a pull request to your branch later.

Reply:
https://github.com/kolbe/tidb/pull/1 @kolbe PTAL

Reply:
@winkyao I added some comments on your PR. I re-wrote my own interpretation of how I see this working and pushed it to my branch of my fork, please take a look.

Basically I added a kind of unsettling amount of additional stuff to create a custom error in config.go, then use that in the case of failed config validation. That enables checking for the specific error type in main.go, and if configStrict is not enabled, it'll get the string from the warning and keep that until ater logging has been set up.

This is kind of a lot of extra work and adds some ugly things, but the whole idea behind strict config checking (and the --config-strict option) is that this is a _temporary_ situation until we make --config-strict behavior the default (and hopefully only!) behavior of TiDB Server in the future. I think at that point we'd simply remove all this extra stuff with the custom error and strange handling in the server.

Thoughts?